Fix for: focused input in dialog is scrolled out of viewport when soft keyboard appears

engineering-this opened the pull request:
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [X] Unit tests
- [X] Manual tests

## What changes did you make?

This is early PR with proposal of solution.

### Current solution

When soft keyboard appears on touch devices it causes 'resize' event on which we were recalculating center position of fixed dialog. After that browser tried to scroll into view of input but it took wrong position, probably before moving it.
Such complex way of centering dialog is unnecessary, as dialog can't be dragged without mouse.

I have reworked how a dialog is styled:
- Dialogs wrapper now has `position: fixed` with `width` and `height` set to 100%.
- When the dialog is shown `document.body` gets a class with 'overflow:hidden', and style `padding-right` calculated from width of scrollbar.
- In IE the dialog is centered via `absolute` position calculated by script.
- In any other browser the dialog is centered by only CSS.
- When the dialog is is dragged it becomes `absolute` positioned.

After changes we won't adjust dialogs position once opened on the mobile devices, it will be positioned only by CSS and the way it's scrolled into view of focused inputs is only relying on browser implementation.

Unfortunately browsers implementation isn't perfect and sometimes it still scrolls it in wrong way, but that's rare. And it should be mentioned that this can happen even with fully static pages, with no js, and everything relative positioned.

### Other solution

Adding `once( 'scroll' ...` inside `focus` listener and calling `scrollIntoView` on currently focused input. This is much simpler, but creates visible flickering, so I dropped this idea for now.

### Todo

- Adding styles to other skins, and testing them.
- Improving way how missing scrollbars are covered with padding, or dropping idea to let page expand.
- Unit tests.
- Cleaning code.

Closes #2395. Closes #3359. Closes #453.

jacekbogdanski commented on plugins/dialog/plugin.js:
```diff
@@ -2107,6 +2128,37 @@ CKEDITOR.DIALOG_STATE_BUSY = 2;
+		scrollBarWidth,
+		paddingRight;
+
+	function handleBodyStyles( dialogOpen ) {
```
Instead of passing `dialogOpen` flag, I would split this function into two separate helpers. This function is used only in two places with strictly passed boolean value, so there is no need for such complication. Also, smaller but verbose functions don't need additional comments. What about:

`addDocumentBodyStyles`
`removeDocumentBodyStyles`

jacekbogdanski commented on plugins/dialog/plugin.js:
```diff
@@ -954,23 +946,41 @@ CKEDITOR.DIALOG_STATE_BUSY = 2;
-					viewSize = win.getViewPaneSize();
 
-			var posX = ( viewSize.width - dialogSize.width ) / 2,
+			if ( !this._.moved && ( !CKEDITOR.env.ie || CKEDITOR.env.edge ) ) {
```
I wonder why you using `flex` for dialog wrapper if the view has not been moved. Isn't better to just recalculate dialog position for the first load? It seems like without this condition dialog works better on iOS (maybe accident, requires additional testing). Not sure if `flex` should be used as a standalone CSS property (outside flexbox layout), especially that it's used with `table` layout - which isn't great also.

engineering-this replied:
I'm using `flex` because it is easiest way to center element inside another element with pure CSS, so we don't need to have JS for that. 

Using `display: flex` on wrapper will make all its children flex items, unless they have defined `position: fixed` or `position: absolute`. Even if dialog outermost element is using table layout to position its children it can be flex item to position itself inside flex wrapper.

engineering-this commented:
@jacekbogdanski  this fix is meant for Android devices, but also affects all modern browsers (in positive way I think). Although it doesn't fix issues with iOS, it doesn't look worse in any way. Mobile browsers in general has some problems with focusing inputs, especially when page is zoomed and I think it works pretty good on iOS.

engineering-this commented:
I've added unit tests, changelog entry, styles for other skins and manuals for each skin.

engineering-this commented on tests/plugins/dialog/positioning.js:
```diff
@@ -0,0 +1,106 @@
+			} );
+		},
+		// When drag starts, dialog becomes centered with `position:absolute`, then it moves together with mouse (#2395).
+		'test dialog move': function() {
```
![screen shot 2018-09-24 at 15 53 02](https://user-images.githubusercontent.com/23366849/45956401-b1384780-c012-11e8-8c2b-47c34fd3c6b4.png)
Passes for me.

jacekbogdanski replied:
Unit test still fails on IE browsers:

![image](https://user-images.githubusercontent.com/10279586/46358280-fee22f00-c666-11e8-9c86-d71b4a703adf.png)

engineering-this commented on tests/plugins/dialog/positioning.js:
```diff
@@ -0,0 +1,106 @@
+		// When dialog is shown body has overflow hidden from class 'cke_dialog_open'
+		// and right padding equal to the scrollbars width. When there is predefined right padding
+		// it should be preserved. All extra styling is removed on dialog hide (#2395).
+		'test body has hidden scrollbars when dialog is opened': function() {
```
Passes for me 😕.

engineering-this replied:
Ok I got it. Tests might fail with small browser window or landscape mode on mobiles. Thats because dialog can be moved only within viewport. I will fix it.

engineering-this commented on plugins/dialog/plugin.js:
```diff
@@ -2107,6 +2128,36 @@ CKEDITOR.DIALOG_STATE_BUSY = 2;
+			// Add right padding, to prevent from resizing elements, when scrollbars are gone.
+			// Don't do anything, when there is any padding on body element, to not break things.
+			body.setStyle( 'padding-right', scrollBarWidth + 'px' );
+			body.addClass( 'cke_body_extra_padding' );
```
Just a reminder. We discussed within our team that it's ok to add prefixed class to body. Also we are removing dynamic padding. It's just not so important visual improvements and we are not interested in putting more energy on that, as there are lots of possible issues following that change.

engineering-this commented:
>@engineering-this referencing to #2413 (comment) - I see your point. However, I'm not sure if we have any real profit from additional condition thus you still have to calculate dialog position for Internet Explorer. After our F2F talk I thought that this code is called every time when the browser window is resized. Although now I see that this point is valid only for IE (correct me if I'm wrong). So we are introducing additional, misleading condition, however, the rest of the code smoothly handles this case.

>Could you elaborate if you see any additional benefit (like performance etc. ) which I don't see? Note that layout have to still be recalculated if dialog has been moved.

The fix seems to work pretty good on mobiles, but since we don't have any way to detect mobile I don't see other way. On the other hand CSS role is to position elements. And centering an element in a view with absolute position calculated with JS and adjusting it on every resize event is an overkill. I don't know if there is benchmark that could compare rendering speed in our cases, but I bet just dropping resize listener for non IE browsers is good for performance.

Theoretically we could completely drop listening for resize, as it is possible to center element in IE8+ with CSS only too, but that's a bit hacky, and requires additional wrapper. We could either keep 'flex' for modern browsers and use hax for IE or use haxy IE code for everything.

```
<div style="display:table;width:100%;height:100%;">First container.
  <div style="display:table-cell;vertical-align:middle;">Second container.
    <div style="margin-left:auto;margin-right:auto;">This is centered element.</div>
  </div>
</div>
```

I don't like one thing tho: we are changing dom structure which leads risk of breaking things. So I'm for keeping old logic for IE as it works there perfectly fine.

engineering-this commented:
Changes:

- Removed more parts of code, mostly code that didn't do anything.
- Updated test due to plugin changes.
- Fixed test failing for small browser window.

@jacekbogdanski I have one concern. How do you think we should deal with following case:

1. User drags dialog, dialog it becomes absolutely positioned.
2. User resizes window, so dialog is outside of viewport.

Should we leave it as it is now (scrollbars appears once dialog is outside of viewport). Or move dialog into view on resize?

jacekbogdanski commented on plugins/dialog/plugin.js:
```diff
@@ -848,10 +843,13 @@ CKEDITOR.DIALOG_STATE_BUSY = 2;
-				element.setStyle( 'display', 'block' );
+			} else {
+				var display = ( !CKEDITOR.env.ie && CKEDITOR.env.edge ) ? 'block' : 'flex';
+				element.setStyle( 'display', display );
```
You could just pass it directly into function execution e.g.
```javascript
element.setStyle( 'display', !CKEDITOR.env.ie && CKEDITOR.env.edge ? 'block' : 'flex' );
```

and additional condition brackets are redundant.

engineering-this commented:
- Changed base to major.
- Changed dialog positioning from pixel values to % of parent.
- Fixed issues with IE.

Dialog still keeps its position when closing and opening again or resizing, but values are now in percentages, so resizing moves a bit dialog, keeping proportions.

engineering-this commented:
I restored resize listener for every browser and changed positioning back to pixel values.

When resizing window dialogs `left` and `top` values should be same fraction of free space (view pane size - dialog size). Outputting prencetage value would calculate the dialogs position based on a container size (which matches view pane). In such situation resizing window results in changing ratio of `left` and `right` or `top` and `bottom`.

Tried to do same calculations with CSS `calc`, but it doesn't work with viewport smaller than dialog.

See how it works right now:

![dialog-correct-resize](https://user-images.githubusercontent.com/23366849/46606181-36842780-cafc-11e8-9322-8bbbade0a198.gif)

Compared how it works without resize listener:

![dialog-incorrect-resize](https://user-images.githubusercontent.com/23366849/46606186-3a17ae80-cafc-11e8-8b33-50bfcaef55d0.gif)

engineering-this commented on plugins/dialog/plugin.js:
```diff
@@ -781,50 +786,54 @@ CKEDITOR.DIALOG_STATE_BUSY = 2;
-
-			// Save the current position.
-			this._.position = { x: x, y: y };
+		move: ( function() {
```
I don't want to pollute dialog object which is accessible from global scope with variable that's needed only inside one method. Therefore I'm moving variable to the upper scope.

engineering-this commented:
>The viewport is zoomed in into an inserted link so close that you have to zoom out to see anything.
>
>I wonder if this issue could be fixed with this PR 🤔

Well, that happens also without this fix. I don't think however this is part of original issue. I doubt we can easily fix that, as this is another part of mobile browser behaviour.

jacekbogdanski commented on tests/plugins/dialog/positioning.js:
```diff
@@ -0,0 +1,177 @@
+			} );
+		},
+
+		'test dialog resize': function() {
```
This test is also failing on Chrome, Android (when runned from the Dashboard)
![image](https://user-images.githubusercontent.com/10279586/47265016-cb861800-d521-11e8-90b2-f082677a92a1.png)
IE8:
![image](https://user-images.githubusercontent.com/10279586/47265099-f1f88300-d522-11e8-9ee1-66c5084c881e.png)

engineering-this commented:
I will rebase once `master` is ready for commits.

jacekbogdanski commented:
Rebased into major.

engineering-this commented:
>In a classic text direction, input is focused at the beginning and a typed text is visible. However, when using an editor in RTL, the caret is hidden outside of the viewport. The issue probably should be fixed by a smaller zoom on focus, so the whole input is visible rather than its beginning. Not sure if it's possible to control?

I can't reproduce it on Android. If typing makes caret visible on iOS I'd leave it as it is.

Mobile browsers decide if it should zoom depending on inputs text size if it's big enough it won't be zoomed, but it varies on various devices. On my phone it doesn't zoom when font is set to 18px or higher. So I tried to hack, by changing font size before focus, and restoring it on focus. This doesn't work, because zoom happens even later. Wrapping it with setTimeout is nogo, it causes nasty flickering, and we don't have any way to detect mobiles, so it would be visible on desktops.

Changing `font-size` only on `input` element is not good idea, as it would look ugly when `input` would have much bigger font than anything else inside dialog.

If we want further improvements we should consider adding new ticket for restyling dialog on mobiles. Whole dialog could be styled with `@media` for mobiles, with bigger fonts, maybe a fullscreen dialogs, we could prevent all the zooming and scrolling, but such big rework exceeds scope of a bugfix, hence it should be extracted as follow up issue.

jacekbogdanski commented on tests/plugins/dialog/manual/skins/kama.md:
```diff
@@ -0,0 +1,5 @@
+@bender-tags: 2395, bug, 4.11.2
```
Update version tag among the tests.

engineering-this commented:
Rebased, but I don't see mentioned issues.

engineering-this commented:
I propose to postpone further reviews until #3034 is merged. It feels like changes to sample on #3034 together with this PR complete each other.

jacekbogdanski commented:
https://github.com/ckeditor/ckeditor-dev/pull/3034 is already merged and there are some merge conflicts. Please, rebase this PR to the latest upstream.

engineering-this commented:
Rebased on the latest `major`.

f1ames commented:
Taking over this review (@jacekbogdanski feel free to drop any comment, I will also ping you for some final check).

engineering-this commented:
>The general manual test (tests/plugins/dialog/manual/inputfocus) seems to not work as expected:

It looks like things got broken from rebase 😭

engineering-this commented:
>1. The general manual test (tests/plugins/dialog/manual/inputfocus) seems to not work as expected:

It looks like rebase fixed this.

>2. Similar issues happens in moonolisa skin test (tests/plugins/dialog/manual/skins/moonolisa). Tabs are misplaced and viewport calculations are affected by left testing panel (doesn't happen with other skins 🤔):

Same as above.

>3. When dialog is opened in too small window, closed, and the opened in resized window, it cannot be dragged vertically on first drag:

Some improvements to dialog relative positioning fixes the issue.

>4. Some dialogs can be resized more than the viewport (while it was not previously possible):

It can be reproduced with any dialog when it wasn't previously dragged.

https://github.com/ckeditor/ckeditor-dev/blob/5c406b3d83c8d1f7cd32a541b20f790a3ce2bdbe/plugins/dialog/plugin.js#L2087-L2093

`getPosition` returns an empty object for dialog which was never repositioned. This is fixed by setting default values  `position.x=0` and `position.y=0`.

>5. There is also an issue with resizing dialogs over viewport boundaries, which causes scrollbars to appear:

Caused by #3144, retrieving values of inline styles instead of `element.getSize` seems to fix it.

>Unit tests
- IE11: somehow these changes make unstable test more likely to fail, however it still sometimes fails on `major`, it is reported here: #3147
Everything else is green for me with most recent changes.

>Skins
I see there were some modifications needed in skin CSS files. This means this PR may be incompatible with 3rd-party skins. @engineering-this could you take a look how 3rd-party skins behaves with this PR? If anything breaks badly we will have to think if something can be improved. Still it is targeted to major so breaking changes are acceptable.

I've reverted some improvements to styling, so dialog tab list is positioned absolutely again. It doesn't affect how editor works in any way, it's just bad practice. All styles brought by this PR are now moved to `plugins/dialog/styles/dialog.css`. I've tested all the custom skins available on https://ckeditor.com/cke4/addons/skins/all
One extra little change was required, but now each one of them works correctly with dialog changes.

>The only thing I'm wondering about is auto-centering dialog on window resize. It's still better than before - you could just resize the window in a way to hide a dialog entirely from viewport.
>One solution is a proposed approach of auto-centering, the other one would be to wait until dialog touches viewport boundaries on resize and only then reposition it to be fully visible.
>I don't have a strong on any particular solution and the one already in place is probably simpler.

Keeping position of dialog when resizing based on ratio of free space around dialog has one nice benefit: when you increase window size again dialog will move back to it's initial position.

f1ames commented:
Rebased on the latest `major`.

f1ames commented:
@engineering-this can you also check if this PR solves any of issues mentioned in https://github.com/ckeditor/ckeditor-dev/pull/2396 (in the `Closes`) section?

engineering-this commented:
>When dialog is dragged, then resized and then window browser is resized it jumps on the first resize (looks like centering).

There is plenty of operations that has to be done on every window resize when dialog moved. We could do this purely in CSS, this way there is no visible lag/delay at all:
```
left: calc( ( freeSpaceWidth - dialogWidth ) * widthRatio )
```
Value should be updated on dialog resize/reposition, but `calc` is supported since IE9, so we would have to keep old logic there just for one browser.

engineering-this commented:
>@engineering-this can you also check if this PR solves any of issues mentioned in #2396 (in the Closes) section?

#748 - not fixed
#895 - failing
#1429 - not fixed
#2420 - thats a new feature so nope

engineering-this commented:
I've found and fixed another issue that was introduced with this PR.

When window was resized to very small size so that scrollbars appear dialog was moved to left by few pixels. It was caused by that we used viewport size, but body with `overflow:hidden` can't have scrollbars. So host for scrollbars is dialog container. To fix this container clientHeight/Width has to be used instead of window viewpane size.

f1ames commented:
Rebased onto latest `major`.

engineering-this commented on plugins/dialog/plugin.js:
```diff
@@ -727,35 +741,36 @@ CKEDITOR.DIALOG_STATE_BUSY = 2;
-					this._.position.x = CKEDITOR.document.getWindow().getViewPaneSize().width - this._.contentSize.width - parseInt( this._.element.getFirst().getStyle( 'right' ), 10 );
+			// Update dialog position when dimension get changed in RTL.
+			if ( this._.editor.lang.dir == 'rtl' && this._.position ) {
+				var containerWidth = dialog.parts.dialog.getParent().getClientSize().width;
```
This fixes some of the issues 👍

engineering-this commented on plugins/dialog/plugin.js:
```diff
@@ -704,6 +715,9 @@ CKEDITOR.DIALOG_STATE_BUSY = 2;
 	function resizeWithWindow( dialog ) {
 		var win = CKEDITOR.document.getWindow();
 		function resizeHandler() {
+			if ( window.prev ) {
```
Leftover. I use window properties for debugging 😃

engineering-this commented:
>It doesn't always position dialog correctly on Android (see video below):
>
>GIF-190827_170618.mp4.zip
>
>Looks like sometimes it works and sometimes input focused by default is still invisible (which was the initial #2395 report).

I don't think we can improve it even further with our current solution. By moving all the positioning logic to CSS we give away responsibility to the browsers. And we can't debug and fix it any further on our side.

Solution that would work 100% of times without flickering will consume way more of our work time. It might be even impossible to do so without mobile browser sniffing.

Still I think current solution is good enough, and we should proceed with it.

After adding following meta tag to test, things seems to work and look even better, we should advocate it somewhere in our docs.
```html
<meta name="viewport" content="width=device-width,initial-scale=1">
```
>Input manual test is not scrollable on iOS when run from a dashboard in a sequence (opening in a new tab works fine) - so it's not possible to perform any testing. Not sure how we can solve this.

It's the same in other tests eg. `/tests/plugins/autoembed/manual/autoembed`. I can't scroll down to `Pass` and `Fail` buttons. Looks like iOS Safari doesn't like our CSS in tests, so this is upstream to bender.js.

f1ames commented:
Rebased onto latest `major`.

f1ames commented:
I see this PR also fixes #453 🎉
